engine/lib/phx/src/audio/audio.rs

```diff
@@ -1,227 +1,83 @@
 
-#[no_mangle]
-pub unsafe extern "C" fn Audio_Update() {
-    FMODCALL(FMOD_System_Update(this.handle));
-    Audio_SetListenerPos(this.autoPos, this.autoVel, this.autoFwd, this.autoUp);
+            let sound_handle = self
+                .audio_manager
+                .play(sound.sound_data().clone())
+                .expect("Cannot play sound");
 
-    let mut soundsToRemove: Vec<usize> = Vec::new();
-    for (i, sound) in this.playingSounds.iter().enumerate() {
-        /* TODO : Refine the API to make this less awkward */
-        if !Sound_IsFreed(&mut **sound) && Sound_IsPlaying(&mut **sound) as i32 != 0 {
-            Sound_Update(&mut **sound);
-        } else {
-            soundsToRemove.push(i);
+            sound.set_sound_handle(sound_handle);
         }
     }
-    for i in soundsToRemove.iter().rev() {
-        this.playingSounds.swap_remove(*i);
-    }
 
-    for sound in this.freeingSounds.iter() {
-        Audio_DeallocSound(*sound);
+    pub fn set_listener_pos(&mut self, pos: &Vec3, rot: &Quat) {
+        self.listener
+            .set_position(*pos, Tween::default()) // TODO: set correct tween
+            .expect("Cannot set listener position");
+        self.listener
+            .set_orientation([rot.x, rot.y, rot.z, rot.w], Tween::default()) // TODO: set correct tween
+            .expect("Cannot set listener position");
     }
-    this.freeingSounds.clear();
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Audio_GetLoadedCount() -> i32 {
-    let size: u32 = StrMap_GetSize(&mut *this.descMap);
-    // Assert(size <= INT32_MAX);
-    return size as i32;
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Audio_GetPlayingCount() -> i32 {
-    this.playingSounds.len() as i32
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Audio_GetTotalCount() -> i32 {
-    let size: u32 = MemPool_GetSize(&mut *this.soundPool);
-    // Assert(size <= INT32_MAX);
-    size as i32
-}
 
-#[no_mangle]
-pub unsafe extern "C" fn Audio_GetHandle() -> *mut libc::c_void {
-    this.handle as *mut _
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Audio_AllocSoundDesc(name: *const libc::c_char) -> *mut SoundDesc {
-    let mut desc: *mut SoundDesc = StrMap_Get(&mut *this.descMap, name) as *mut SoundDesc;
-    if desc.is_null() {
-        desc = MemNewZero!(SoundDesc);
-        StrMap_Set(&mut *this.descMap, name, desc as *mut _);
+    pub fn get_loaded_count(&self) -> u64 {
+        self.audio_manager.num_sounds() as u64
     }
-    desc
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Audio_DeallocSoundDesc(desc: *mut SoundDesc) {
-    StrMap_Remove(&mut *this.descMap, (*desc).name);
-    MemFree(desc as *const _);
-}
 
-#[no_mangle]
-pub unsafe extern "C" fn Audio_AllocSound() -> *mut Sound {
-    MemPool_Alloc(&mut *this.soundPool) as *mut Sound
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn Audio_DeallocSound(sound: *mut Sound) {
-    MemPool_Dealloc(&mut *this.soundPool, sound as *mut _);
-}
+    pub fn get_playing_count(&self) -> u64 {
+        0 // FIXME
```

Comment:
Was this missed, or being addressed later?

Reply:
Honestly I didn't find where to take this number from yet. :)

Reply:
I found a potential way to implement this but it is not trivial. Each Sound object has method is_playing() so we can go over all sounds and find who is playing. But at the moment we don't have common register of sounds.

So we can keep all sounds in Audio object. In this case we will have to use Rc/RefCell (or even Arc/Mutex) to share sounds between Rust and Lua code.

I think for now it's too complicated to implement. We can take a look at it later.